commons/zenoh-keyexpr/src/key_expr/canon.rs

```diff
@@ -12,114 +12,103 @@
-            if in_big_wild {
-                match chunk {
-                    [SINGLE_WILD] | b"$*" => {
-                        writer.write(b"/*");
-                    }
-                    DOUBLE_WILD => {}
-                    _ => {
-                        writer.write(b"/**/");
-                        writer.write(chunk);
-                        in_big_wild = false;
+            _ => {
+                if double_wild && &bytes[index..] != b"**" {
+                    bytes[written..written + 3].copy_from_slice(b"**/");
+                    written += 3;
+                    double_wild = false;
+                }
+                let mut write_start = index;
+                loop {
+                    match bytes.get(index) {
+                        Some(b'/') => {
+                            index += 1;
+                            bytes.copy_within(write_start..index, written);
+                            written += index - write_start;
+                            break;
+                        }
+                        Some(b'$') if matches!(bytes.get(index + 1..index + 4), Some(b"*$*")) => {
+                            index += 2;
+                            bytes.copy_within(write_start..index, written);
+                            written += index - write_start;
+                            let skip = bytes[index + 4..]
+                                .windows(2)
+                                .take_while(|s| s == b"$*")
+                                .count();
+                            index += (1 + skip) * 2;
+                            write_start = index;
+                        }
+                        Some(_) => index += 1,
+                        None => {
+                            bytes.copy_within(write_start..index, written);
+                            written += index - write_start;
+                            return written;
+                        }
                     }
                 }
-            } else if chunk == DOUBLE_WILD {
-                in_big_wild = true;
-            } else {
-                writer.write_byte(DELIMITER);
-                writer.write(if chunk == b"$*" { b"*" } else { chunk });
-            }
-        }
-        if in_big_wild {
-            if writer.len != 0 {
-                writer.write_byte(DELIMITER);
             }
-            writer.write(DOUBLE_WILD)
-        }
-        *self = unsafe {
-            str::from_utf8_unchecked_mut(slice::from_raw_parts_mut(writer.ptr, writer.len))
         }
     }
 }
 
+impl Canonizable for &mut str {
+    fn canonize(&mut self) {
+        // SAFETY: canonize leave an UTF8 string within the returned length,
+        // and remaining garbage bytes are zeroed
+        let bytes = unsafe { self.as_bytes_mut() };
+        let length = canonize(bytes);
+        // SAFETY: canonized length is lesser than or equal to the original length
+        unsafe { bytes.get_unchecked_mut(length..) }.fill(b'\0');
```

Comment:
@Mallets I don't know how much performance of canonization are critical, but do you think I can remove this unsafe?

Reply:
Yes, go ahead.